### PR TITLE
Add DbSet support for shared type entity types

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #9914

### CODE REVIEW

File: src/EFCore/ChangeTracking/Internal/EntityTypeFinder.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.ChangeTracking.Internal
+{
+    public class EntityTypeFinder : IEntityTypeFinder
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/Internal/EntityTypeFinder.cs]: Move this to a separate PR as a stretch stretch goal

File: src/EFCore/Internal/InternalDbSet.cs [hunk 2]
```diff
@@ -61,8 +63,11 @@ private IEntityType EntityType
                     return _entityType;
                 }
 
-                _entityType = _context.Model.FindEntityType(typeof(TEntity));
+                _entityType = !string.IsNullOrEmpty(_entityTypeName)
+                    ? _context.Model.FindEntityType(_entityTypeName)
+                    : _context.Model.FindEntityType(typeof(TEntity));
 
+                // TODO: improve exception messages for shared entity types
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/InternalDbSet.cs]: Comment not relevant anymore

**Comment by smitpatel** [comment]
Updated.

File: src/EFCore/DbContext.cs [hunk 3]
```diff
@@ -237,13 +237,38 @@ object IDbSetCache.GetOrAddSet(IDbSetSource source, Type type)
...
+            return set;
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        [EntityFrameworkInternal]
```
Comment (AndriySvyryd) [inline src/EFCore/DbContext.cs]: I don't think the attribute makes any difference since this is an explicit interface implementation

File: src/EFCore/Internal/InternalDbSet.cs [hunk 4]
```diff
@@ -61,7 +63,9 @@ private IEntityType EntityType
                     return _entityType;
                 }
 
-                _entityType = _context.Model.FindEntityType(typeof(TEntity));
+                _entityType = !string.IsNullOrEmpty(_entityTypeName)
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/InternalDbSet.cs]: Just compare to `null`

File: test/EFCore.Tests/DbContextTest.cs [hunk 5]
```diff
@@ -727,7 +727,7 @@ public async Task It_throws_object_disposed_exception(bool async)
             await Assert.ThrowsAsync<ObjectDisposedException>(() => context.FindAsync(typeof(Random), 77).AsTask());
 
             var methodCount = typeof(DbContext).GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly).Count();
-            var expectedMethodCount = 42;
+            var expectedMethodCount = 42 + 1;
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/DbContextTest.cs]: 😆
